### ISSUE TITLE
combineLatest doesn't not work correctly with zero input streams

### ISSUE THREAD

**Issue by trotyl** [ReactiveX/rxjs#5039]
## Bug Report

**Current Behavior**
A clear and concise description of the behavior.

`combineLatest` takes an array of input streams, namely, **zero** or more input streams. But in the case of **zero** input streams, it hangs infinitely instead of emit an empty Array/Tuple (`[]`).

**Reproduction**
- REPL or Repo link:

https://stackblitz.com/edit/rxjs-a961ph?devtoolsheight=60

```js
// Promise

Promise.all([Promise.resolve(1), Promise.resolve(2)]) // => [1, 2]
Promise.all([Promise.resolve(1)]) // => [1]
Promise.all([]) // => []

// RxJS

combineLatest([of(1), of(2)])  // => [1, 2]
combineLatest([of(1)]) // => [1]
combineLatest([]) // Never produce a value
```

**Expected behavior**

```js
combineLatest([]) // => []
```

As per the doc:

> Combines multiple Observables to create an Observable whose values are calculated from the latest values of each of its input Observables.

RxJS should follow the conventional definition of **each**/**every** in algorithms, for a zero length sequence, its `.every(criteria)` is always `true`.

**Environment**
- Runtime: [e.g. Node vUnrelavant, Chrome vUnrelavant]
- RxJS version: 6.5.2
- (If bug is related) Loader, build configuration: [e.g webpack, angular-cli version, config]

**Possible Solution**

**Additional context/Screenshots**
Add any other context about the problem here. If applicable, add screenshots to help explain.

Relates to https://github.com/ReactiveX/rxjs/issues/3189

**Comment by cartant** [comment]
`[]` is the input stream though. And it is empty.

Arrays are valid `ObservableInput` values and `combineLatest` accepts rest parameters: `combineLatest(of(1), of(2))`.

I'm not saying that this isn't confusing and cannot be improved, though.

**Comment by trotyl** [comment]
> `[]` is the input stream though. And it is empty.

@cartant Changing to `combineLatest()` would have the same result:

https://stackblitz.com/edit/rxjs-qu7wtj?file=index.ts

Is there any reason for **"no arguments at all"** being considered an input stream?

**Comment by cartant** [comment]
Yeah. I'm not saying it's not inconsistent. The inconsistency is one reason I wrote [this](https://github.com/cartant/rxjs-etc/blob/master/source/observable/combineLatestArray.ts). It's just not the sort of 'bug' that can be fixed without being deemed a breaking change.

**Comment by dobesv** [comment]
I guess the workaround here is to use `rxjs-etc` `combineLatestArray` if you want to pass an array to `combineLatest` that could be empty.

**Comment by benlesh** [comment]
Hi @trotyl — thank you for taking the time to open this issue and for sharing your thoughtful feedback with the RxJS project. I’m sorry it took us so long to respond. Given how much time has passed and the lack of recent activity, I’m going to close this for now so the tracker reflects the work we’re likely to act on. If this is still relevant on a current RxJS release, please feel welcome to open a new issue with updated context and a reproduction. Thank you again for your contribution and patience.

— Ben
